codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoderTest.java

```diff
@@ -587,8 +600,21 @@ public void pushPromiseReadAfterGoAwayShouldAllowFramesForStreamCreatedByLocalEn
         verify(listener).onPushPromiseRead(eq(ctx), anyInt(), anyInt(), any(Http2Headers.class), anyInt());
     }
 
+    @Test
+    public void pushPromiseReadForUnknownStreamThatMayHaveExistedShouldBeIgnored() throws Exception {
```

Comment:
I don't think we should be ignoring these, we should be resetting. The [spec states](https://http2.github.io/http2-spec/#rfc.section.5.1) (bottom of the closed state description):

> An endpoint might receive a PUSH_PROMISE frame after it sends RST_STREAM. PUSH_PROMISE causes a stream to become "reserved" even if the associated stream has been reset. Therefore, a RST_STREAM is needed to close an unwanted promised stream.

Reply:
@bryce-anderson I agree but I also don't know how to fix this. In https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java#L766 we require the parent stream to be non-null. Do you have an idea in mind in how to make this compliant with the spec?

Reply:
Do we need to use `shouldIgnoreHeadersOrDataFrame` for a PUSH_PROMISE? From a brief read it doesn't seem appropriate since it's using the parent streams ID in the `streamCreatedAfterGoAwaySent(streamId)` method but we should be considering the promised stream id as the GOAWAY frame we would have sent carries the last _peer-initiated_ stream id to be processed.

To fix the behavior can we do the following: ignore the frame if `streamCreatedAfterGoAwaySent(promisedStreamId) == true`, otherwise send a RST_STREAM if the parent stream is `null`, otherwise continue on with the existing code path?

Reply:
+1, running the tests now.

Reply:
Also, I suppose we should check if the parent stream ever could have existed and if the promised stream does get sent a reset, we need to make sure we 'observe' the promised stream id via `incrementExpectedStreamId(streamId);` which unfortunately is in the `DefaultHttp2ConnectionreservePushStream method`. A lot of protocol checks in there apply to the promise frame even if the parent has already been reset. Maybe the right thing to do is throw a `streamError` in there instead of a protocol error if the parent stream doesn't exist (which if it's valid would be because it's reset)?

Reply:
@bryce-anderson let me know if the latest commit is in line with what you were thinking. I wasn't 100% sure what affect throwing a stream error has but it seems to eventually send a RST_STREAM. is that accurate?

https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java#L704